azkaban-common/src/main/java/azkaban/executor/ExecutorManager.java

```diff
@@ -153,189 +263,328 @@ public long getLastCleanerThreadCheckTime() {
 
   private void loadRunningFlows() throws ExecutorManagerException {
     runningFlows.putAll(executorLoader.fetchActiveFlows());
+    // Finalize all flows which were running on an executor which is now
+    // inactive
+    for (Pair<ExecutionReference, ExecutableFlow> pair : runningFlows.values()) {
+      if (!activeExecutors.contains(pair.getFirst().getExecutor())) {
+        finalizeFlows(pair.getSecond());
+      }
+    }
   }
 
+  /*
+   * load queued flows i.e with active_execution_reference and not assigned to
+   * any executor
+   */
+  private void loadQueuedFlows() throws ExecutorManagerException {
+    queuedFlows.addAll(executorLoader.fetchQueuedFlows());
+    for (Pair<ExecutionReference, ExecutableFlow> ref : queuedFlows) {
+      queuedFlowMap.put(ref.getSecond().getExecutionId(), ref);
+    }
+  }
+
+  /**
+   * Gets a list of all the active (running, non-dispatched) executions for a
+   * given project and flow {@inheritDoc}
+   *
+   * @see azkaban.executor.ExecutorManagerAdapter#getRunningFlows(int,
+   *      java.lang.String)
+   */
   @Override
   public List<Integer> getRunningFlows(int projectId, String flowId) {
     ArrayList<Integer> executionIds = new ArrayList<Integer>();
-    for (Pair<ExecutionReference, ExecutableFlow> ref : runningFlows.values()) {
+    getRunningFlowsHelper(projectId, flowId, executionIds,
+      queuedFlowMap.values());
+    getRunningFlowsHelper(projectId, flowId, executionIds,
+      runningFlows.values());
+    return executionIds;
+  }
+
+  /* Helper method for getRunningFlows */
+  private void getRunningFlowsHelper(int projectId, String flowId,
```

Comment:
I was thinking that it might be better for readability if we do
1. this method returns List<Integer>
2. the calling method calls  executionIds.addAll(retval)

Wondering if people have thoughts/preferences to this